Geo charts

cryogenian opened the pull request:
This is not ready for review, but if someone could take a quick look it would be really helpful. 

+ Introduced `GeoChartType` (similar with `ChartType`, `FormInputType`) 
+ Unified api of `FooType` modules
+ Added three card types: Setup Heatmap, Setup Marker, Show Geo 
+ Added one more case for `EvalState` 

How it works: 
+ Setups use the same approach as chart setups with `Package`, lenses, projection etc
+ Heatmap setup works with `Free` package builder 
+ Marker is much more like `Parallel` setup because they both have dynamic dimensions
+ Geo port is effectful function from `Leaflet` (that's a map) to `Array Layer`, effects are living in `Aff`, because most things in **purescript-leaflet** works with `MonadEff` and `Alternative`
+ On eval of setups this effectful function is constructed and put to eval state. 
+ On initialization of **purescript-halogen-leaflet** child `Leaflet` is put to eval state. Then model is updated to run eval once again. Then layers are constructed and put to eval state. 
+ Component eats layers from eval state.

The most worrying thing for me is `Aff` in port. I'd like to change it to something existentially hidden and general but this won't work, right?
```purescript
newtype GeoPortE m = GeoPortE (Leaflet -> m (Array Layer))
data GeoPort 
hide :: forall m. Alternative m => MonadEff _ m => GeoPortE m -> GeoPort 
hide = unsafeCoerce 

reveal :: forall r. GeoPort -> (forall m. Alternative m => MonadEff _ m => GeoPortE m -> r) -> r
reveal p fn = fn $ unsafeCoerce p
```

@natefaubion @garyb What do you think about such eval call? I'm not totally sure that this is correct way of doing things.

beckyconning commented:
Cool! I'm reading through now : ).

beckyconning commented on src/SlamData/Workspace/Card/CardType/ChartType.purs:
```diff
@@ -167,8 +167,8 @@ chartLightIconSrc = case _ of
 
-chartDarkIconSrc ∷ ChartType → String
-chartDarkIconSrc = case _ of
+darkIconSrc ∷ ChartType → String
```
I like this renaming re qualified imports.

beckyconning commented on src/SlamData/Workspace/Card/Eval.purs:
```diff
@@ -202,5 +213,11 @@ extractResourceVar dm = case SM.toUnfoldable (Port.filterResources dm) of
 extractResource ∷ ∀ m. MonadThrow CE.CardError m ⇒ Port.DataMap → m (Port.Resource)
```
This is an aside but what is the relationship between `extractResource` here and `extractResource` here?

https://github.com/cryogenian/slamdata/blob/e2748068f47097900f92af87ac64c77d83597ea0/src/SlamData/Workspace/Card/Port.purs#L174

natefaubion replied:
There's an `extractResource` in `CEM` which is an assertion, and throws an error if it's missing. It uses the one in `Port`.

natefaubion replied:
But why were these moved?

natefaubion replied:
Oh ignore me, I thought that lived somewhere else :)

cryogenian replied:
I didn't touch them.

beckyconning replied:
Not related to this pr but perhaps eventually this could be renamed to `extractSingletonResource` and the other one `extractFirstResource` or something.

beckyconning replied:
Should I make comments like these elsewhere? I'm aware they don't really fit into the review of a PR however I'm not sure where else to make them and its nice to do it with the source right there... perhaps I should just write up a ticket or slack document and link to the lines in question.

cryogenian replied:
I think we could use `MonadThrow` and `Alternative` constraints everywhere. Such helpers become almost redundant with this

beckyconning commented on src/SlamData/Workspace/Card/Eval.purs:
```diff
@@ -202,5 +213,11 @@ extractResourceVar dm = case SM.toUnfoldable (Port.filterResources dm) of
 
-tapResource ∷ ∀ m. MonadThrow CE.CardError m ⇒ (Port.Resource → m Port.Port) → Port.DataMap → m Port.Out
-tapResource f dm = map (_ × dm) (f =<< extractResource dm)
+tapResource
```
**Aside**

Might be clearer without this but I might think this because I don't understand the use of the word "tap" here.

My current understanding is essentially that:

* This expression is needed because some `eval`s which take resources return `Port` instead of `Out`. 
* This is a top level value here because its used multiple times in this file. 

This took me like 20 minutes to work out but that might be my own shortcoming rather than the source's.

cryogenian replied:
You know I'm not very good in English, for me `tapResource` is just pure phenomenon w/o any associations :)

natefaubion replied:
`tapResource` feeds the resource into the function, but doesn't let the function change it. It's for evals that need the input resource, but don't create a new resource.

beckyconning commented on src/SlamData/Workspace/Card/Geo/Eval.purs:
```diff
@@ -0,0 +1,44 @@
+  → Port.Resource
+  → m Port.Port
+eval gcPort resource = do
+  let path = resource ^. Port._filePath
```
**Aside**

I find it kinda confusing that this `_filePath` `Lens` is in `Port`. It reads like it gets the file path from a port rather than a resource. 

But then there there is a `Traversal` called `_filePath` in `Resource` so...

https://github.com/slamdata/slamdata/blob/master/src/SlamData/FileSystem/Resource.purs#L321

beckyconning replied:
I only mention this because I felt like I had to check that that this wasn't a weird mistake where `resource` and `gcPort` had been swapped.

beckyconning replied:
And with something called `all` on the next line and a traversal called the same in `Resource` I managed to get my head in a spin lol.

cryogenian replied:
Oh! You're right! I'm not sure how to workaround this, but this is really confusing

beckyconning commented on src/SlamData/Workspace/Card/Geo/Eval.purs:
```diff
@@ -0,0 +1,44 @@
+  case evalState of
+    Just (ES.Geo st) → do
+      layers ← case st.leaflet of
+        Nothing → pure st.layers
```
Could you describe the cases where `st.leaflet` is `Nothing` and layers is not `[]`?

cryogenian replied:
`st.leaflet` is `Nothing` when leaflet isn't initialized.
`st.layers` isn't `[]` when leaflet is initialized and all calculations are done.

beckyconning replied:
So this line always returns pure []?

cryogenian replied:
Maybe you are right, I'll doublecheck

beckyconning commented on src/SlamData/Workspace/Card/Geo/Eval.purs:
```diff
@@ -0,0 +1,44 @@
+        Just l → liftAff $ build l
+      put $ Just $ ES.Geo { build, layers, leaflet: st.leaflet }
+    _ → put $ Just $ ES.Geo { leaflet: Nothing, build, layers: [ ] }
+  get >>= traceAnyA
```
I'm sure you're still using this but just in case.

beckyconning commented on src/SlamData/Workspace/Card/Geo/Model.purs:
```diff
@@ -0,0 +1,147 @@
+import Test.StrongCheck.Gen as Gen
+
+type ModelR =
+  { osmURI ∷ URIRef
```
I know what Open Street Maps is but not everyone will. Perhaps add a comment?

cryogenian replied:
:+1"

beckyconning commented on src/SlamData/Workspace/Card/Geo/Model.purs:
```diff
@@ -0,0 +1,147 @@
+initialModel ∷ Model
+initialModel = Nothing
+
+eqR ∷ ModelR → ModelR → Boolean
```
Zero judgement just wondering how you decided to do this rather than having a newtype and deriving equality?

cryogenian replied:
This is for consistency with charts. Charts has been working this way for ages, I think it was designed even before record equality could be derived.

beckyconning replied:
Cool thanks!

beckyconning commented on src/SlamData/Workspace/Card/Geo/Model.purs:
```diff
@@ -0,0 +1,147 @@
+
+-- | OSMURI is a template that can have "{" and "}".
+-- | This is invalid URI and we need to handle it somehow
+onURIRef ∷ (String → String) → URIRef → URIRef
```
Bookmark for me to carry on reading in the morning : ).

beckyconning commented on src/SlamData/Workspace/Card/Geo/Model.purs:
```diff
@@ -0,0 +1,147 @@
+-- | OSMURI is a template that can have "{" and "}".
+-- | This is invalid URI and we need to handle it somehow
+onURIRef ∷ (String → String) → URIRef → URIRef
+onURIRef f = case _ of
```
I believe this applies the provided function to all the strings in a `URIRef` however I had to read thirty lines to work that out.

If my belief is correct I think I'd prefer a name like `uriRefStringApply` and a comment like 

```
-- Applies the given function to every `String` in `Either URI RelativeRef`.
-- This is needed here to encode and decode `OSMURIRef`s as they are templates containing `{` and `}`.

For a moment I thought it would be cool to have a new data type called like `TemplateURI` where every part could be `Either Template` but that would be way more work than its worth for just encoding / decoding.

It might be useful to rename this to something do something like the following?

```
newtype OSMURIRef = OSMURIRef URIRef
derive instance newtypeOSMURIRef :: Newtype OSMURIRef

toValidURIRef :: OSMURIRef -> URIRef
toValidURIRef = 
  uriRefStringApply encodeURIComponent

fromValidURIRef :: URIRef -> OSMURIRef
fromValidURIRef = 
  uriRefStringApply decodeURIComponent
```

cryogenian replied:
Yeah, I thought about it too. It will ends with same code but one more type. As for names maybe `everywhereInURIRef` seemed good to me, but again maybe comment is enough.

beckyconning commented on src/SlamData/Workspace/Card/InsertableCardType.purs:
```diff
@@ -126,7 +130,10 @@ inputsFor card =
 outputFor ∷ InsertableCardType → Maybe InsertableCardIOType
 outputFor card =
-  Utils.singletonValue Nothing (const Nothing) (map snd $ Array.filter (eq card ∘ fst) outputs)
+  Utils.singletonValue
```
: D

beckyconning replied:
Maybe this could be used for `extractResourceVar` too.

beckyconning commented on src/SlamData/Workspace/Card/InsertableCardType.purs:
```diff
@@ -298,25 +309,23 @@ aAn s =
-    , action
-    , "."
-    ]
+reason io card = fold
```
good spot!

cryogenian commented:
Depends on slamdata/purescript-halogen-leaflet#2 

Icons: 
![icons-1](https://cloud.githubusercontent.com/assets/10245930/26685880/fffc7146-46f3-11e7-89da-ee3f880c8802.png)
![icons-2](https://cloud.githubusercontent.com/assets/10245930/26685881/00953476-46f4-11e7-92ec-63e186bdee31.png)

+ Added comment for `onURIRef`
+ Added comment for `osmURI`
+ Switched to `for_` in `Geo.Eval` 
+ Added layer control for marker map

@beckyconning Please review

cryogenian commented:
@toastal Could you review/modify icons, please?

toastal commented:
@cryogenian Currently they match the style of the card icons so that's probably okay. I'd stick to that for now.

cryogenian commented:
/cc @garyb

cryogenian commented:
Fixes #1826 

+ Added model/state for osm uri

cryogenian commented:
Starting work on #1827 from here

garyb commented on src/Test/StrongCheck/Data/String.purs:
```diff
@@ -0,0 +1,37 @@
+      Gen.elements "a"
+      $ foldMap (pure ∘ Str.fromCharArray ∘ pure)
+      $ Str.toCharArray "bcdefghijklmnopqrstuvwxyz"
+    pure $ acc <> letter
```
You could use stuff from `Data.String.Gen` / `Data.Char.Gen` for this :)

garyb commented on src/Utils.purs:
```diff
@@ -28,6 +28,9 @@ import Data.String as S
 import SqlSquared.Signature.Ident (printIdent)
 
+
+foreign import debugTime ∷ ∀ a. String → (Unit → a) → a
```
If we're going to keep this in the source, perhaps it's worth adding a `Warn` constraint so it's not forgotten and left in production code.

cryogenian commented:
+ Added `Warn` constraint
+ Fixed tests (e.g. switched to `genAlphaLowercase`) 

I think this is ready for review again 🙂

cryogenian commented:
+ Added wrapper for `debugTime`

cryogenian commented:
:tada: Thanks a lot for reviewing this!
